src/CorePolicy.cc

```diff
@@ -0,0 +1,201 @@
+/* Copyright (c) 2015-2017 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <thread>
+#include "PerfUtils/TimeTrace.h"
+#include "PerfUtils/Util.h"
+#include "PerfUtils/Cycles.h"
+#include "CoreArbiter/CoreArbiterClient.h"
+
+#include "Arachne.h"
+#include "CorePolicy.h"
+#include "PerfStats.h"
+#include "CoreArbiter/ArbiterClientShim.h"
+
+using PerfUtils::Cycles;
+
+/**
+  * Arachne will attempt to increase the number of cores if the idle core
+  * fraction (computed as idlePercentage * numSharedCores) is less than this
+  * number.
+  */
+double maxIdleCoreFraction = 0.1;
+
+/**
+  * Arachne will attempt to increase the number of cores if the load factor
+  * increase beyond this threshold.
+  */
+double loadFactorThreshold = 1.0;
+
+/**
+  * Only here to satisfy compiling benchmark.
+  */
+double maxUtilization = 0.9;
+
+/**
+  * Save the core fraction at which we ramped up based on load factor, so we
+  * can decide whether to ramp down.
+  */
+double *utilizationThresholds;
+
+/**
+  * The difference in load, expressed as a fraction of a core, between a
+  * ramp-down threshold and the corresponding ramp-up threshold.
+  */
+double idleCoreFractionHysteresis = 0.2;
+
+/**
+  * The percentage of slots which are occupied that will prevent reducing the

```

Comment:
What slots?

Also, grammatically incorrect, and not clear which way the inequality works. How about this, and instead?

If at least this fraction of <what kind of?> slots are occupied, we will not reduce the number of cores.